src/backend/distributed/connection/remote_commands.c

```diff
@@ -395,106 +447,159 @@ GetRemoteCommandResult(MultiConnection *connection, bool raiseInterrupts)
+ *
+ * Returns false if PQputCopyEnd() failed, true otherwise.
+ */
+bool
+PutRemoteCopyEnd(MultiConnection *connection, const char *errormsg)
+{
+	PGconn *pgConn = connection->pgConn;
+	int copyState = 0;
 
-		/* stop writing if all data has been sent, or there was none to send */
-		if (rc == 0)
-		{
-			break;
-		}
+	if (PQstatus(pgConn) != CONNECTION_OK)
+	{
+		return false;
+	}
 
-		/* if sending failed, there's nothing more we can do */
-		if (rc == -1)
-		{
-			failed = true;
-			break;
-		}
+	Assert(PQisnonblocking(pgConn));
 
-		/* this means we have to wait for data to go out */
-		Assert(rc == 1);
+	copyState = PQputCopyEnd(pgConn, errormsg);
 
-#if (PG_VERSION_NUM >= 100000)
-		rc = WaitLatchOrSocket(MyLatch, waitFlags | WL_SOCKET_WRITEABLE, socket, 0,
-							   PG_WAIT_EXTENSION);
-#else
-		rc = WaitLatchOrSocket(MyLatch, waitFlags | WL_SOCKET_WRITEABLE, socket, 0);
-#endif
+	if (copyState == 1)
+	{
+		/* successful */
+		return true;
+	}
+	else if (copyState == -1)
+	{
+		return false;
+	}
+	else
+	{
+		bool allowInterrupts = true;
+		return FinishConnectionIO(connection, allowInterrupts);
+	}
+}
 
-		if (rc & WL_POSTMASTER_DEATH)
-		{
-			ereport(ERROR, (errmsg("postmaster was shut down, exiting")));
-		}
 
-		if (rc & WL_LATCH_SET)
-		{
-			/* if allowed raise errors */
-			if (raiseInterrupts)
-			{
-				CHECK_FOR_INTERRUPTS();
-			}
+/*
+ * FinishConnectionIO performs pending IO for the connection, while accepting
+ * interrupts.
+ *
+ * See GetRemoteCommandResult() for documentation of interrupt handling
+ * behaviour.
+ *
+ * Returns true if IO was successfully completed, false otherwise.
+ */
+static bool
+FinishConnectionIO(MultiConnection *connection, bool raiseInterrupts)
+{
+	PGconn *pgConn = connection->pgConn;
+	int socket = PQsocket(pgConn);
 
-			/*
-			 * If raising errors allowed, or called within in a section with
-			 * interrupts held, return NULL instead, and mark the transaction
-			 * as failed.
-			 */
-			if (InterruptHoldoffCount > 0 && (QueryCancelPending || ProcDiePending))
-			{
-				connection->remoteTransaction.transactionFailed = true;
-				failed = true;
-				break;
-			}
-		}
+	Assert(pgConn);
+	Assert(PQisnonblocking(pgConn));
+
+	if (raiseInterrupts)
+	{
+		CHECK_FOR_INTERRUPTS();
 	}
 
-	/* wait for the result of the command to come in */
-	while (!failed)
+	/* perform the necessary IO */
+	while (true)
 	{
+		int sendStatus = 0;
 		int rc = 0;
+		int waitFlags = WL_POSTMASTER_DEATH | WL_LATCH_SET;
 
-		ResetLatch(MyLatch);
+		/* try to send all pending data */
+		sendStatus = PQflush(pgConn);
+
+		/* if sending failed, there's nothing more we can do */
+		if (sendStatus == -1)
+		{
+			return false;
+		}
+		else if (sendStatus == 1)
+		{
+			waitFlags |= WL_SOCKET_WRITEABLE;
+		}
 
 		/* if reading fails, there's not much we can do */
 		if (PQconsumeInput(pgConn) == 0)
 		{
-			failed = true;
-			break;
+			return false;
+		}
+		if (PQisBusy(pgConn))
+		{
+			waitFlags |= WL_SOCKET_READABLE;
```

Comment:
You just made me realise I overlooked this :). We should always check for reads even if the socket is not busy, since the server could be sending notices and get blocked on them, meaning it won't accepts writes until we read.